### PR TITLE
feat(storage): enables location credentials provider

### PR DESCRIPTION
**Pull request by israx**
#### Description of changes

The following PR enables the location credentials provider on the `resolveS3ConfigAndInput` helper. This helper is used by all of the storage APIs.

##### Main changes:

- Removed `identityId` assertion in the `resolveS3ConfigAndInput` helper  — doubled check with @AllanZhengYP already. The assertion of the `identityId` is not needed by the underlying service APIs. 
- The `credentialsProvider` in the `resolveS3ConfigAndInput` function is asserting for a `deprecated` input or a callback `path` if and only if a `locationCredentialsProvider` is passed in the Storage option inputs.

#### Description of how you validated changes
- unit tests

#### Checklist

- [x] PR description included
- [x] `yarn test` passes
- [x] Unit Tests are [changed or added](https://github.com/aws-amplify/amplify-js/blob/main/CONTRIBUTING.md#steps-towards-contributions)
- [ ] Relevant documentation is changed or added (and PR referenced)

#### Checklist for repo maintainers

- [ ] Verify E2E tests for existing workflows are working as expected or add E2E tests for newly added workflows
- [ ] New source file paths included in this PR have been added to CODEOWNERS, if appropriate

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: packages/storage/__tests__/providers/s3/apis/utils/resolveS3ConfigAndInput.test.ts [hunk 1]
```diff
@@ -76,13 +81,11 @@ describe('resolveS3ConfigAndInput', () => {
 		}
 	});
 
-	it('should throw if identityId is not available', async () => {
+	it('should not throw if identityId is not available', async () => {
```
Comment (israx) [inline packages/storage/__tests__/providers/s3/apis/utils/resolveS3ConfigAndInput.test.ts:84]: The `resolveS3ConfingAndInput` helper is not asserting for the identityId anymore

File: packages/storage/src/providers/s3/utils/resolveS3ConfigAndInput.ts [hunk 2]
```diff
@@ -38,13 +59,13 @@ interface ResolvedS3ConfigAndInput {
...
 	amplify: AmplifyClassV6,
 	apiOptions?: S3ApiOptions,
+	input?: DeprecatedStorageInput | CallbackPathStorageInput,
 ): Promise<ResolvedS3ConfigAndInput> => {
 	/**
 	 * IdentityId is always cached in memory so we can safely make calls here. It
 	 * should be stable even for unauthenticated users, regardless of credentials.
 	 */
 	const { identityId } = await amplify.Auth.fetchAuthSession();
-	assertValidationError(!!identityId, StorageValidationErrorCode.NoIdentityId);
```
Comment (AllanZhengYP) [inline packages/storage/src/providers/s3/utils/resolveS3ConfigAndInput.ts:47]: We confirmed that this validation is extra by validating every reference of identity id. The identity id is modeled as optional through the codebase. When the identity id is required, like prefix resolver, we have the same assertion throwing same error.

**Review by AllanZhengYP — COMMENTED** [review]
It looks good to me, just a couple nits.

This approach makes sense to me. If we cannot prevent complexity, I'd rather contain the complexity inside this function and refactor it later, than scatter it in multiple modules.

File: packages/storage/src/providers/s3/utils/resolveS3ConfigAndInput.ts [hunk 3]
```diff
@@ -38,13 +59,13 @@ interface ResolvedS3ConfigAndInput {
 export const resolveS3ConfigAndInput = async (
 	amplify: AmplifyClassV6,
 	apiOptions?: S3ApiOptions,
+	input?: DeprecatedStorageInput | CallbackPathStorageInput,
```
[the remark this answers is not in the document]
Comment (israx) [reply packages/storage/src/providers/s3/utils/resolveS3ConfigAndInput.ts:62]: would like to leave it as `input`. I have other functions that make reference to the `input` param

File: packages/storage/src/providers/s3/types/inputs.ts [hunk 4]
```diff
@@ -35,19 +35,24 @@ import {
 	UploadDataOptionsWithPath,
 } from '../types';
 
+import { CommonOptions } from './options';
+
 // TODO: support use accelerate endpoint option
```
Comment (AllanZhengYP) [inline packages/storage/src/providers/s3/types/inputs.ts:40]: Since we don't have people asking for this option, it makes sense to keep leaving it out of the `options`.

**Review by AllanZhengYP — DISMISSED** [review]

File: packages/storage/src/providers/s3/utils/resolveS3ConfigAndInput.ts (hunk 2 above)
Comment (ashika112) [reply packages/storage/src/providers/s3/utils/resolveS3ConfigAndInput.ts:47]: @AllanZhengYP will that be the case in Path callback as well? if they are using identityId and it is not preset they will have an error thrown out?

File: packages/storage/src/providers/s3/utils/resolveS3ConfigAndInput.ts [hunk 5]
```diff
@@ -101,3 +128,70 @@ export const resolveS3ConfigAndInput = async (
...
+			typeof (input as StorageOperationInputWithPath).path === 'function') ||
+		((input as StorageCopyInputWithPath)?.destination?.path &&
+			typeof (input as StorageCopyInputWithPath).destination?.path ===
+				'function') ||
+		((input as StorageCopyInputWithPath)?.source?.path &&
+			typeof (input as StorageCopyInputWithPath).source?.path === 'function')
+	);
+};
+
+const isDeprecatedInput = (
```
Comment (ashika112) [inline packages/storage/src/providers/s3/utils/resolveS3ConfigAndInput.ts:152]: name is misleading since callbackPath is not deprecated

File: packages/storage/src/providers/s3/apis/internal/remove.ts [hunk 6]
```diff
@@ -25,7 +25,7 @@ export const remove = async (
 ): Promise<RemoveOutput | RemoveWithPathOutput> => {
 	const { options = {} } = input ?? {};
 	const { s3Config, keyPrefix, bucket, identityId } =
-		await resolveS3ConfigAndInput(amplify, options);
+		await resolveS3ConfigAndInput(amplify, options, input as RemoveInput);
```
Comment (ashika112) [inline packages/storage/src/providers/s3/apis/internal/remove.ts:28]: why are we typecasting here? shouldn't the input take both?

File: packages/storage/src/providers/s3/utils/resolveS3ConfigAndInput.ts [hunk 7]
```diff
@@ -101,3 +128,70 @@ export const resolveS3ConfigAndInput = async (
...
+	return (
+		isInputWithKey(input) ||
+		isInputWithPrefix(input) ||
+		isInputWithCopySourceOrDestination(input)
+	);
+};
+const assertStorageInput = (
+	input?: DeprecatedStorageInput | CallbackPathStorageInput,
+) => {
+	if (isDeprecatedInput(input) || isInputWithCallbackPath(input)) {
```
Comment (ashika112) [inline packages/storage/src/providers/s3/utils/resolveS3ConfigAndInput.ts:164]: why not use assertValidationError here?

File: packages/storage/src/types/inputs.ts [hunk 8]
```diff
@@ -85,12 +85,14 @@ export type StorageUploadDataInputWithPath<Options> =
 export interface StorageCopyInputWithKey<
 	SourceOptions extends StorageOptions,
 	DestinationOptions extends StorageOptions,
-> {
+	Options,
+> extends StorageOperationOptionsInput<Options> {
```
Comment (ashika112) [inline packages/storage/src/types/inputs.ts:89]: can we just add `options: Options` here

File: packages/storage/src/types/inputs.ts [hunk 9]
```diff
@@ -85,12 +85,14 @@ export type StorageUploadDataInputWithPath<Options> =
...
 	source: SourceOptions;
 	destination: DestinationOptions;
 }
 
-export interface StorageCopyInputWithPath {
+export interface StorageCopyInputWithPath<Options = unknown>
```
Comment (ashika112) [inline packages/storage/src/types/inputs.ts:94]: why set it to unknown?

File: packages/storage/src/providers/s3/utils/resolveS3ConfigAndInput.ts [hunk 10]
```diff
@@ -101,3 +128,70 @@ export const resolveS3ConfigAndInput = async (
...
+		throw new StorageError({
+			name: INVALID_STORAGE_INPUT,
+			message: 'The storage input needs to pass a string path.',
+		});
+	}
+};
+
+const isInputWithKey = (
```
Comment (ashika112) [inline packages/storage/src/providers/s3/utils/resolveS3ConfigAndInput.ts:172]: nit: should we move these to separate utils and test them out separately?

File: packages/storage/__tests__/providers/s3/apis/utils/resolveS3ConfigAndInput.test.ts (hunk 1 above)
Comment (ashika112) [reply packages/storage/__tests__/providers/s3/apis/utils/resolveS3ConfigAndInput.test.ts:84]: while this might not be needed, if customers are already relyign on this message for something, would removing it cause a breaking change for them?

File: packages/storage/src/providers/s3/apis/internal/remove.ts (hunk 6 above)
Comment (jimblanc) [reply packages/storage/src/providers/s3/apis/internal/remove.ts:28]: +1, can we get by without the cast?

File: packages/storage/src/providers/s3/utils/resolveS3ConfigAndInput.ts [hunk 11]
```diff
@@ -101,3 +128,45 @@ export const resolveS3ConfigAndInput = async (
...
+	return !!options?.locationCredentialsProvider;
+};
+
+const isInputWithCallbackPath = (input?: CallbackPathStorageInput) => {
+	return (
+		((input as StorageOperationInputWithPath)?.path &&
+			typeof (input as StorageOperationInputWithPath).path === 'function') ||
+		((input as StorageCopyInputWithPath)?.destination?.path &&
+			typeof (input as StorageCopyInputWithPath).destination?.path ===
+				'function') ||
```
[the remark this answers is not in the document]
Comment (jimblanc) [reply packages/storage/src/providers/s3/utils/resolveS3ConfigAndInput.ts:146]: Agreed, this did end up getting implemented? See the feedback was resolved

**Review by jimblanc — COMMENTED** [review]
Overall looks good! Agree with some of the feedback from other reviewers (left comments)

File: packages/storage/src/providers/s3/utils/resolveS3ConfigAndInput.ts (hunk 2 above)
Comment (AllanZhengYP) [reply packages/storage/src/providers/s3/utils/resolveS3ConfigAndInput.ts:47]: Oh this is a good point. The path callback's input identity id is optional. However, since we used to validate identity id here, the path callbacks's input identity id is never actually optional although it's typed as optional. By removing the validation here, we probably should validate identity id in `validateStorageOptionalInput()`:https://github.com/aws-amplify/amplify-js/blob/58b86bb8930061d406fe0f03a7d335c4fb40060a/packages/storage/src/providers/s3/utils/validateStorageOperationInput.ts#L25

File: packages/storage/src/types/inputs.ts (hunk 8 above)
Comment (AllanZhengYP) [reply packages/storage/src/types/inputs.ts:89]: +1 I think this is the better approach.

File: packages/storage/src/providers/s3/utils/resolveS3ConfigAndInput.ts (hunk 2 above)
Comment (ashika112) [reply packages/storage/src/providers/s3/utils/resolveS3ConfigAndInput.ts:47]: yes we made it named param since we might need to make way for additional values here but not sure why we made it optional. Now that we are here more , i think it shouldn't have been optional 😓

File: packages/storage/src/providers/s3/utils/resolveS3ConfigAndInput.ts [hunk 12]
```diff
@@ -23,6 +36,14 @@ interface ResolvedS3ConfigAndInput {
...
 	identityId?: string;
 }
+export type DeprecatedStorageInput =
+	| StorageOperationInputWithKey
+	| StorageOperationInputWithPrefix
+	| CopyInput;
+
+export type CallbackPathStorageInput =
+	| StorageOperationInputWithPath
+	| StorageCopyInputWithPath;
```
Comment (Samaritan1011001) [inline packages/storage/src/providers/s3/utils/resolveS3ConfigAndInput.ts:46]: Should this be `CopyWithPathInput`?

File: packages/storage/src/providers/s3/utils/resolveS3ConfigAndInput.ts [hunk 13]
```diff
@@ -101,3 +128,70 @@ export const resolveS3ConfigAndInput = async (
...
+		isInputWithCopySourceOrDestination(input)
+	);
+};
+const assertStorageInput = (
+	input?: DeprecatedStorageInput | CallbackPathStorageInput,
+) => {
+	if (isDeprecatedInput(input) || isInputWithCallbackPath(input)) {
+		throw new StorageError({
+			name: INVALID_STORAGE_INPUT,
+			message: 'The storage input needs to pass a string path.',
```
Comment (Samaritan1011001) [inline packages/storage/src/providers/s3/utils/resolveS3ConfigAndInput.ts:167]: Nit: Since it can be either string or function, this message can be updated a bit or have it generic with a specific recoverySuggestion.

**Review by Samaritan1011001 — COMMENTED** [review]
Looks good, few questions. Thanks Israel!

File: packages/storage/__tests__/providers/s3/apis/utils/resolveS3ConfigAndInput.test.ts (hunk 1 above)
Comment (AllanZhengYP) [reply packages/storage/__tests__/providers/s3/apis/utils/resolveS3ConfigAndInput.test.ts:84]: @ashika112 I don't think removing the validation here itself is a breaking change, unless they are using Storage APIs intentionally to check for identity ID. However, it's not the intended usage of the Storage APIs.

Comment (ashika112) [reply packages/storage/__tests__/providers/s3/apis/utils/resolveS3ConfigAndInput.test.ts:84]: Oh i mean , if customer lets say is specifically finding this error to avoid making a call right now, wouldn't it break? this is propogated all the way to customer. Again coming from path callback perspective, but if this check is moved to path validation i think it is fine there. but wondering if there is any other use case like that

File: packages/storage/src/providers/s3/utils/resolveS3ConfigAndInput.ts (hunk 7 above)
Comment (israx) [reply packages/storage/src/providers/s3/utils/resolveS3ConfigAndInput.ts:164]: The `assertValidationError` uses a map and enum, expanding any of those will increase the bundle size. That is why I would like to avoid using it

File: packages/storage/src/types/inputs.ts (hunk 8 above)
Comment (israx) [reply packages/storage/src/types/inputs.ts:89]: The `StorageOperationOptionsInput` is already used by other inputs. Want to keep consistency

Screenshot 2024-07-19 at 3 04 03 PM

File: packages/storage/src/providers/s3/utils/resolveS3ConfigAndInput.ts (hunk 12 above)
Comment (israx) [reply packages/storage/src/providers/s3/utils/resolveS3ConfigAndInput.ts:46]: The 	`StorageCopyInputWithPath` type is a type that we are alredy exposing

File: packages/storage/src/types/inputs.ts (hunk 9 above)
Comment (israx) [reply packages/storage/src/types/inputs.ts:94]: The `StorageCopyInputWithPath` doesn't extend the `CommonOptions` as the other Inputs, hence it doesn't have any constrains. The `unknown` type is a placeholder that allows to use the `StorageCopyInputWithPath` type without providing a generic type.

File: packages/storage/src/providers/s3/utils/resolveS3ConfigAndInput.ts (hunk 10 above)
Comment (israx) [reply packages/storage/src/providers/s3/utils/resolveS3ConfigAndInput.ts:174]: These helpers are mainly used to test `deprecated` inputs, and I added a test that will test all the possible permutations.

File: packages/storage/__tests__/providers/s3/apis/utils/resolveS3ConfigAndInput.test.ts [hunk 14]
```diff
@@ -214,4 +217,74 @@ describe('resolveS3ConfigAndInput', () => {
...
+
+			it.each(testCases)('should throw when input is %s', async input => {
+				const { s3Config } = await resolveS3ConfigAndInput(
+					Amplify,
+					{ locationCredentialsProvider: mockLocationCredentialsProvider },
+					input,
+				);
+				if (typeof s3Config.credentials === 'function') {
+					await expect(s3Config.credentials()).rejects.toThrow(
+						expect.objectContaining({
```
Comment (ashika112) [inline packages/storage/__tests__/providers/s3/apis/utils/resolveS3ConfigAndInput.test.ts:277]: nit:we could just check for name here since if we make changes to message and recovery in future we need to come update it here. that is not better pratice

File: packages/storage/src/providers/s3/apis/downloadData.ts [hunk 15]
```diff
@@ -115,7 +115,11 @@ const downloadDataJob =
 	> => {
 		const { options: downloadDataOptions } = downloadDataInput;
 		const { bucket, keyPrefix, s3Config, identityId } =
-			await resolveS3ConfigAndInput(Amplify, downloadDataOptions);
+			await resolveS3ConfigAndInput(
```
Comment (ashika112) [inline packages/storage/src/providers/s3/apis/downloadData.ts:118]: General comment: can be sep PR - should we just destructure the options inside `resolveS3ConfigAndInput` since we are passing whole input now

File: packages/storage/src/providers/s3/utils/resolveS3ConfigAndInput.ts (hunk 7 above)
Comment (ashika112) [reply packages/storage/src/providers/s3/utils/resolveS3ConfigAndInput.ts:164]: Yes but any validation we already perform it through the assertValidation why change the pattern here and now?

File: packages/storage/src/providers/s3/utils/resolveS3ConfigAndInput.ts (hunk 10 above)
Comment (ashika112) [reply packages/storage/src/providers/s3/utils/resolveS3ConfigAndInput.ts:174]: Hmmm i notice the test so the test right now only covers the happy path cases what if the value passed is invalid is not covered. I dont mind this staying here but we should cover all branches in tests so we dont introduce bugs in future

File: packages/storage/src/types/inputs.ts (hunk 8 above)
Comment (ashika112) [reply packages/storage/src/types/inputs.ts:89]: hmm also copy with key shouldn't need to add an options right?

File: packages/storage/src/types/inputs.ts (hunk 9 above)
Comment (ashika112) [reply packages/storage/src/types/inputs.ts:94]: Neither does `copywithkey` sorry im confused. I still dont see the need here, pass a generic and use it. if need arises we can do this.

**Review by ashika112 — DISMISSED** [review]
My other comments are non-blocking and ok to approve here. Lets try to follow up the other ones or refute them. thanks

File: packages/storage/src/providers/s3/utils/resolveS3ConfigAndInput.ts (hunk 10 above)
Comment (AllanZhengYP) [reply packages/storage/src/providers/s3/utils/resolveS3ConfigAndInput.ts:174]: I don't think we need to export util functions just to unit test it. However it makes sense to have the unit test of `resolveS3ConfigAndInput` to cover these code paths dealing with deprecated inputs.

File: packages/storage/src/types/inputs.ts (hunk 9 above)
Comment (israx) [reply packages/storage/src/types/inputs.ts:93]: removed the extra type

File: packages/storage/src/providers/s3/utils/resolveS3ConfigAndInput.ts (hunk 10 above)
Comment (israx) [reply packages/storage/src/providers/s3/utils/resolveS3ConfigAndInput.ts:174]: @ashika112 — The `resolveS3ConfigAndInput` test is also testing negative cases, specially when `locationCredentialsProvider` is passed and the input is invalid

File: packages/storage/src/providers/s3/utils/resolveS3ConfigAndInput.ts [hunk 16]
```diff
@@ -23,6 +50,30 @@ interface ResolvedS3ConfigAndInput {
...
+	| ListPaginateInput
+	| GetPropertiesInput
+	| CopyInput;
+
+export type CallbackPathStorageInput =
+	| StorageOperationInputWithPath
+	| StorageCopyInputWithPath;
+
+type StorageInput =
+	| DeprecatedStorageInput
```
Comment (ashika112) [inline packages/storage/src/providers/s3/utils/resolveS3ConfigAndInput.ts:69]: I think types here can be simplified. DMed u on this

File: packages/storage/src/providers/s3/utils/resolveS3ConfigAndInput.ts (hunk 10 above)
Comment (ashika112) [reply packages/storage/src/providers/s3/utils/resolveS3ConfigAndInput.ts:206]: Replied on the tests on my question

File: packages/storage/__tests__/providers/s3/apis/utils/resolveS3ConfigAndInput.test.ts [hunk 17]
```diff
@@ -214,4 +217,74 @@ describe('resolveS3ConfigAndInput', () => {
...
+			if (typeof s3Config.credentials === 'function') {
+				const result = await s3Config.credentials();
+				expect(mockLocationCredentialsProvider).toHaveBeenCalled();
+				expect(result).toEqual(credentials);
+			} else {
+				throw new Error('Expect credentials to be a function');
+			}
+		});
+
+		describe('with deprecated or callback paths as inputs', () => {
```
[the remark this answers is not in the document]
Comment (israx) [reply packages/storage/__tests__/providers/s3/apis/utils/resolveS3ConfigAndInput.test.ts:249]: `path` is independent when `locationCredentialsProvider` is called.

**Review by ashika112 — APPROVED** [review]

File: packages/storage/__tests__/providers/s3/apis/utils/resolveS3ConfigAndInput.test.ts (hunk 17 above)
[the remark this answers is not in the document]
Comment (ashika112) [reply packages/storage/__tests__/providers/s3/apis/utils/resolveS3ConfigAndInput.test.ts:249]: yes but it should not throw here tho right? i think we should have a test exercising that route

**Review by ashika112 — APPROVED** [review]

**Review by AllanZhengYP — APPROVED** [review]
